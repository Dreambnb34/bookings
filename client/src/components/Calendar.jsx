import React from 'react';
import DateCell from './Date';

class Calendar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      oneSelected: false,
      dateSelected: undefined,
    };

    this.updateCalendar = this.updateCalendar.bind(this);
  }

  firstDay() {
    const { year, month } = this.props;
    return new Date(year, month, 1).getDay();
  }

  daysInMonth() {
    const { year, month } = this.props;
    return 32 - new Date(year, month, 32).getDate();
  }

  renderHeader() {
    const month = this.props.month;
    const MONTHS = [
      'January',
      'February',
      'March',
      'April',
      'May',
      'June',
      'July',
      'August',
      'September',
      'October',
      'November',
      'December',
    ];
    const monthName = MONTHS[month];
    return monthName;
  }

  updateCalendar(boolean) {
    this.setState({
      oneSelected: JSON.parse(boolean),
    });
  }

  updateDateSelected(key) {
    this.setState({
      dateSelected: key,
    });
  }

  renderDates() {
    const daysInMonth = this.daysInMonth();
    const firstDay = this.firstDay();
    const numOfWeeks = [0, 1, 2, 3, 4, 5];
    const numOfDays = [0, 1, 2, 3, 4, 5, 6];

    let days = 1;

    const weeks = numOfWeeks.map((week, indW) => {
      let singleWeek = [];

      numOfDays.forEach((day, indD) => {
        if (week === 0 && day < firstDay) {
          singleWeek.push(
            <DateCell isEmpty={true} key={'' + indW + indD} date={day} />,
          );
        } else if (days <= daysInMonth) {
          singleWeek.push(
            <DateCell
              available={true}
              selected={false}
              dateSelected={
                this.state.dateSelected ? this.state.dateSelected : 32
              }
              key={'' + indW + indD}
              date={days}
              updateCalendar={this.updateCalendar}
              updateDateSelected={this.updateDateSelected.bind(this)}
            />,
          );
          days += 1;
        } else {
          singleWeek.push(
            <DateCell isEmpty={true} key={'' + indW + indD} date={days} />,
          );
        }
      });
      return <tr key={indW}>{singleWeek}</tr>;
    });

    return [weeks];
  }

  render() {
    let dates = this.renderDates();
    let header = this.renderHeader();
    return (
      <div className="calendar-container">
        <h3 className="calendar-header">{header}</h3>
        <table>
          <th>Su</th>
          <th>Mo</th>
          <th>Tu</th>
          <th>We</th>
          <th>Th</th>
          <th>Fr</th>
          <th>Sa</th>
          <tbody>{dates}</tbody>
        </table>
      </div>
    );
  }
}

export default Calendar;
