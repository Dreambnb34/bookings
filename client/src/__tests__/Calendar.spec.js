import React from 'react';
import { render } from 'react-testing-library';
import Calendar from '../components/Calendar';

const mockData = {
  month: 10,
  year: 2018,
  leftMount: true,
};

it('is rendering out a wrapper div holding a table', () => {
  const { getByTestId, debug } = render(<Calendar month={9} year={2018} />);
  const calendarContainer = getByTestId('calendar-container');
  expect(getByTestId('calendar-container')).toBeDefined();
  //debug(calendarContainer);
});

it('should have a table and th element with text content of "Su', () => {
  const { getByTestId } = render(<Calendar month={9} year={2018} />);

  expect(getByTestId('calendar-table-th-Su')).toHaveTextContent('Su');
});

it('should render out a date cell for Noverber 11', () => {
  const { getByTestId } = render(<Calendar month={10} year={2018} />);
  expect(getByTestId('available-unselected-calendar-day')).toHaveTextContent(
    11,
  );
});

it('should render a div with the role button', () => {
  const { getByTestId } = render(<Calendar month={10} year={2018} />);
  expect(getByTestId('calendar-button')).not.toBeNull();
});
