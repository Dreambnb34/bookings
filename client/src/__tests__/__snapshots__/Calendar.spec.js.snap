// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="calendar-container"
  data-testid="calendar-container"
>
  <div
    data-testid="calendar-button"
    role="button"
  >
    <button
      data-testid="calendar-left-button"
      id="left-button-div"
      onClick={[Function]}
    >
      left
    </button>
    <span />
  </div>
  <h3
    className="calendar-header"
    data-testid="calendar-header"
  >
    November
     
    2018
  </h3>
  <table
    className="calendar-table"
    data-testid="calendar-table"
  >
    <th
      data-testid="calendar-table-th-Su"
    >
      Su
    </th>
    <th
      data-testid="calendar-table-th-Mo"
    >
      Mo
    </th>
    <th
      data-testid="calendar-table-th-Tu"
    >
      Tu
    </th>
    <th
      data-testid="calendar-table-th-We"
    >
      We
    </th>
    <th
      data-testid="calendar-table-th-Th"
    >
      Th
    </th>
    <th
      data-testid="calendar-table-th-Fr"
    >
      Fr
    </th>
    <th
      data-testid="calendar-table-th-Sa"
    >
      Sa
    </th>
    <tbody>
      <tr>
        <td
          className="empty-calendar-day"
        />
        <td
          className="empty-calendar-day"
        />
        <td
          className="empty-calendar-day"
        />
        <td
          className="empty-calendar-day"
        />
        <td
          className="unavailable-calendar-day"
          data-testid="unavailable-calendar-day-1"
        >
          1
        </td>
        <td
          className="unavailable-calendar-day"
          data-testid="unavailable-calendar-day-2"
        >
          2
        </td>
        <td
          className="unavailable-calendar-day"
          data-testid="unavailable-calendar-day-3"
        >
          3
        </td>
      </tr>
      <tr>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-4"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          4
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-5"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          5
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-6"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          6
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-7"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          7
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-8"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          8
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-9"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          9
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-10"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          10
        </td>
      </tr>
      <tr>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-11"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          11
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-12"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          12
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-13"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          13
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-14"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          14
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-15"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          15
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-16"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          16
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-17"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          17
        </td>
      </tr>
      <tr>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-18"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          18
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-19"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          19
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-20"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          20
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-21"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          21
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-22"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          22
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-23"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          23
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-24"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          24
        </td>
      </tr>
      <tr>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-25"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          25
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-26"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          26
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-27"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          27
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-28"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          28
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-29"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          29
        </td>
        <td
          className="available-unselected-calendar-day"
          data-testid="available-unselected-calendar-day-30"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          30
        </td>
        <td
          className="empty-calendar-day"
        />
      </tr>
      <tr>
        <td
          className="empty-calendar-day"
        />
        <td
          className="empty-calendar-day"
        />
        <td
          className="empty-calendar-day"
        />
        <td
          className="empty-calendar-day"
        />
        <td
          className="empty-calendar-day"
        />
        <td
          className="empty-calendar-day"
        />
        <td
          className="empty-calendar-day"
        />
      </tr>
    </tbody>
  </table>
</div>
`;
